src/deploy/functions/build.ts

```diff
@@ -0,0 +1,397 @@
+  requiredAPIs: RequiredApi[];
+  endpoints: Record<string, Endpoint>;
+  params: Param[];
+}
+
+interface RequiredApi {
+  // The API that should be enabled. For Google APIs, this should be a googleapis.com subdomain
+  // (e.g. vision.googleapis.com)
+  api: string;
+
+  // A reason why this codebase requires this API.
+  // Will be considered required for all Extensions codebases. Considered optional for Functions
+  // codebases.
+  reason?: string;
+}
+
+// Defining a StringParam with { param: "FOO" } declares that "{{ params.FOO }}" is a valid
+// Expression<string> elsewhere.
+// Expression<number> is always an int. Float Params, list params, and secret params cannot be used in
+// expressions.
+// `Expression<Foo> == Expression<Foo>` is an Expression<boolean>
+// `Expression<boolean> ? Expression<T> : Expression<T>` is an Expression<T>
+type Expression<T extends string | number | boolean> = string;
+
+function resolveInt(from: number | Expression<number> | null): number {
+  if (from == null) {
+    return 0;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+function resolveString(from: string | Expression<string> | null): string {
+  if (from == null) {
+    return "";
+  }
+  return from;
+}
+
+function resolveBoolean(from: boolean | Expression<boolean> | null): boolean {
+  if (from == null) {
+    return false;
+  } else if (typeof from === "string") {
+    throw new FirebaseError("CEL evaluation of expression '" + from + "' not yet supported");
+  }
+  return from;
+}
+
+// A service account must either:
+// 1. Be a project-relative email that ends with "@" (e.g. database-users@)
+// 2. Be a well-known shorthand (e..g "public" and "private")
+type ServiceAccount = string;
+
+// Trigger definition for arbitrary HTTPS endpoints
+interface HttpsTrigger {
+  // Which service account should be able to trigger this function. No value means "make public
+  // on create and don't do anything on update." For more, see go/cf3-http-access-control
+  invoker?: ServiceAccount | null;
+}
+
+// Trigger definitions for RPCs servers using the HTTP protocol defined at
+// https://firebase.google.com/docs/functions/callable-reference
+// eslint-disable-next-line
+interface CallableTrigger { }
+
+// Trigger definitions for endpoints that should be called as a delegate for other operations.
+// For example, before user login.
+interface BlockingTrigger {
+  eventType: string;
+}
+
+// Trigger definitions for endpoints that listen to CloudEvents emitted by other systems (or legacy
+// Google events for GCF gen 1)
+interface EventTrigger {
+  eventType: string;
+  eventFilters: Record<string, Expression<string>>;
+
+  // whether failed function executions should retry the event execution.
+  // Retries are indefinite, so developers should be sure to add some end condition (e.g. event
+  // age)
+  retry: boolean | Expression<boolean> | null;
```

Comment:
nit: here and elsewhere you could have saved yourself a lot of time by defining `Expression<T>` to be

```
// TOOD RE list
Expression<T extends string | number | boolean> = T extends string ? string | null : T | `{{ ${infer expr} }}` | null
```

And then you can just say

```
retry: Expression<boolean>
```

Or if you want to leave `Expression` as the string type, you could define `Field<T> = T | Expression<T> | null`